### ISSUE TITLE
YOLOv5 segmentation np.uint8 bug

### ISSUE THREAD

**Issue by mqbin** [ultralytics/yolov5#9461]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

I want to try the YOLOv5 segmentation task on my own dataset, but I encounter the following error.

```
Traceback (most recent call last):
  File "/home/mqb/yolov5/segment/train.py", line 676, in <module>
    main(opt)
  File "/home/mqb/yolov5/segment/train.py", line 572, in main
    train(opt.hyp, opt, device, callbacks)
  File "/home/mqb/yolov5/segment/train.py", line 295, in train
    for i, (imgs, targets, paths, _, masks) in pbar:  # batch ------------------------------------------------------
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/home/mqb/yolov5/utils/dataloaders.py", line 170, in __iter__
    yield next(self.iterator)
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 681, in __next__
    data = self._next_data()
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1376, in _next_data
    return self._process_data(data)
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 1402, in _process_data
    data.reraise()
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/_utils.py", line 461, in reraise
    raise exception
TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/_utils/worker.py", line 302, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/mqb/anaconda3/envs/py39/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/mqb/yolov5/utils/segment/dataloaders.py", line 167, in __getitem__
    masks = (torch.from_numpy(masks) if len(masks) else torch.zeros(1 if self.overlap else nl, img.shape[0] //
TypeError: can't convert np.ndarray of type numpy.uint16. The only supported types are: float64, float32, float16, complex64, complex128, int64, int32, int16, int8, uint8, and bool.
```
I customize the dataset image size to be 256*256 and the format is that of the COCO dataset. For training I used the tool [JSON2YOLO](https://github.com/ultralytics/JSON2YOLO) to convert it to YOLO data format and then I got the error, how do I fix this?

The same data works fine for the detection task, but it will be wrong for the segmentation task.
The data format is as follows
```
1 0.253906 0.982422 0.226562 0.978516 0.216797 0.957031 0.246094 0.951172 0.259766 0.96875 0.253906 0.982422
```

### Environment

- torch 1.12.1+cu113 
- python:3.9

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@mqbin your labels look good. The error seems to imply that your labels are in uint16, but they need to be floats for the fractions. 

Are you able to train with COCO128-seg?
```
python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640
```

**Comment by mqbin** [comment]
> @mqbin your labels look good. The error seems to imply that your labels are in uint16, but they need to be floats for the fractions.
> 
> Are you able to train with COCO128-seg?
> 
> ```
> python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640
> ```

@glenn-jocher Yes, I can train with COCO128-seg. Is there something wrong with the format of my dataset?

**Comment by DonkeySmall** [comment]
```
\yolov5\utils\segment\dataloaders.py

def polygons2masks_overlap(img_size, segments, downsample_ratio=1):
    """Return a (640, 640) overlap mask."""
	- masks = np.zeros((img_size[0] // downsample_ratio, img_size[1] // downsample_ratio), dtype=np.uint8)
        + masks = np.zeros((img_size[0] // downsample_ratio, img_size[1] // downsample_ratio), dtype=np.int32)
```

**Comment by glenn-jocher** [comment]
@mqbin yes there's likely an issue with your dataset. You should review the structure and labels vs COCO128-seg to make sure they align.

@DonkeySmall what's the diff you have there?

**Comment by mqbin** [comment]
> ```
> \yolov5\utils\segment\dataloaders.py
> 
> def polygons2masks_overlap(img_size, segments, downsample_ratio=1):
>     """Return a (640, 640) overlap mask."""
> 	- masks = np.zeros((img_size[0] // downsample_ratio, img_size[1] // downsample_ratio), dtype=np.uint8)
>         + masks = np.zeros((img_size[0] // downsample_ratio, img_size[1] // downsample_ratio), dtype=np.int32)
> ```

It works, thanks a lot!

**Comment by mqbin** [comment]
> 

@glenn-jocher , I solved this problem using the method given by @DonkeySmall. But I haven't looked at the code yet,  I don't know why I have this problem. Will this change I make to the code affect the results?

**Comment by glenn-jocher** [comment]
@mqbin I don't know, but in general masks should be uint8 when used for plotting. You're saying there's no issue with COCO128-seg but on your custom dataset you run into an error and the above change resolves this error?

**Comment by glenn-jocher** [comment]
@mqbin ok thank you! I've opened up #9493 to test the fix.

**Comment by Laughing-q** [comment]
@DonkeySmall @mqbin We still can't reproduce this issue, could you please provide us more information on this?

**Comment by mqbin** [comment]
> @DonkeySmall @mqbin We still can't reproduce this issue, could you please provide us more information on this?

@Laughing-q I am using a cell segmentation dataset where each image size is split into 256*256 and there are many more small instances on each image compared to the COCO128 dataset. I am not sure if this is the cause of the problem.

**Comment by glenn-jocher** [comment]
@mqbin can you share the dataset to help us reproduce the problem?

**Comment by DonkeySmall** [comment]
> @DonkeySmall @mqbin We still can't reproduce this issue, could you please provide us more information on this?

Sorry, I can't reproduce this problem anymore, cleaned up my dataset.

**Comment by mqbin** [comment]
> @mqbin can you share the dataset to help us reproduce the problem?

@glenn-jocher I used the [PanNuke](https://jgamper.github.io/PanNukeDataset/) dataset, [this code ](https://github.com/nolancardozo13/pathology_maskrcnn)converts the original PanNuke dataset into the COCO instance segmentation format, then I used JSON2YOLO to convert the format to the YOLO dataset format for instance segmentation training , and the above error occurred.

**Comment by mqbin** [comment]
@glenn-jocher @Laughing-q Can you give me your email address, I will send you a part of the dataset in an email to reproduce the error

**Comment by Laughing-q** [comment]
@mqbin  yeah sure, laughingq@163.com
I just downloaded the datasets, but it'd be better if you send us yours. :)

**Comment by Laughing-q** [comment]
@mqbin @DonkeySmall I've reproduced this issue, so this is about the number of objects. This error occurs when the number of objects is beyond 255 in one image. We'll fix it by the PR Glenn mentioned before.

**Comment by glenn-jocher** [comment]
@mqbin good news 😃! Your original issue may now be fixed ✅ in PR #9493. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
